rpcserver.go

```diff
@@ -2733,6 +2733,33 @@ func (r *rpcServer) AddFederationServer(ctx context.Context,
 
 	serversToAdd := fn.Map(in.Servers, unmarshalUniverseServer)
 
+	// Before we add the server as a federation member, we check that we can
+	// actually connect to it and that it isn't ourselves.
+	for idx := range serversToAdd {
+		server := serversToAdd[idx]
+
+		rpcsLog.Debugf("Attempting to connect to federation server %v",
+			server.HostStr())
+
+		conn, err := ConnectUniverse(server)
+		if err != nil {
+			return nil, fmt.Errorf("error connecting to server "+
+				"%v: %w", server.HostStr(), err)
+		}
+
+		// We don't allow adding ourselves as a federation member.
+		info, err := conn.Info(ctx, &unirpc.InfoRequest{})
+		if err != nil {
+			return nil, fmt.Errorf("error getting info from "+
+				"server %v: %w", server.HostStr(), err)
+		}
+
+		if info.RuntimeId == r.runtimeID {
```

Comment:
Hmm, yeah seems simple enough, wondering if there're other alternatives we should consider 🤔

Reply:
Also fwiw, this isn't the call path that would lead to us adding ourselves, instead it's here: https://github.com/lightninglabs/taproot-assets/blob/7ab8a5760d671d7446d31871b60ba699618abbad/tapcfg/server.go#L180-L201


https://github.com/lightninglabs/taproot-assets/blob/main/universe/auto_syncer.go#L105-L118

This'll actually end up skipping this RPC path.

Reply:
> wondering if there're other alternatives we should consider

I thought about it a bit and I think it's not trivial to find out if we are connecting to ourselves just from the URL/domain name. Because what we have configured as our RPC listener port might be completely different from the domain name/port we are given to connect to (due to load balancers and other things inbetween). So I thought this was the simplest approach.
But I'm open to other approaches.

> Also fwiw, this isn't the call path that would lead to us adding ourselves, instead it's here: 

Yeah, I somehow forgot about that path before. Fixed now.

Reply:
> But I'm open to other approaches.


Hmm, good point about the ports, etc....not sure there's a good option other than this. Only other thing can think of is extending to also support using `brontide` over gRPC, then we'd have the public key we can check against. 

Also if we want a bit more assurance that we won't have to worry about collisions here, then we can sample a `[32]byte` with the CSPRNG. I think we can run with this for now, then circle back to see if was have false positives in the wild (a bit of extra logging can help us detect that). 

In terms of gaming, worst thing I can think of is someone sort of bans a server from connecting to them as they mimic their runtime ID, but then they can already to that by banning the IP itself.

Reply:
Wouldn't we also need to know the public key of the server when using `brontide`? Or is there a variant of the Noise protocol where you learn the other side's long-term keys without also doing SPAKE?

I don't worry about collision too much. If there was a collision between my local daemon and the universe, I would just restart my daemon and the collision would be gone.
And I agree about banning, I don't think we introduce any new assumptions regarding that with this approach.